File: src/main/java/io/koju/autopos/catalog/domain/Category.java
package io.koju.autopos.catalog.domain;

import io.koju.autopos.shared.domain.AuditableEntity;
import org.hibernate.annotations.Cache;
import org.hibernate.annotations.CacheConcurrencyStrategy;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import java.util.Objects;

import static javax.persistence.GenerationType.SEQUENCE;

/**
 * A Category.
 */
@Entity
@Table(name = "category")
@Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
public class Category extends AuditableEntity {

    @Id
    @SequenceGenerator(name = "category_id_seq", sequenceName = "category_id_seq", allocationSize = 1)
    @GeneratedValue(strategy = SEQUENCE, generator = "category_id_seq")
    private Long id;

    @NotNull
    @Size(min = 2, max = 3)
    @Column(name = "short_name", length = 3, nullable = false)
    private String shortName;

    @NotNull
    @Size(min = 3, max = 50)
    @Column(name = "name", length = 50, nullable = false)
    private String name;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getShortName() {
        return shortName;
    }

    public void setShortName(String shortName) {
        this.shortName = shortName;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

}

File: scalang/src/main/scala/com/kapilkoju/autopos/catalog/domain/Category.scala
package com.kapilkoju.autopos.catalog.domain

import javax.persistence._
import javax.validation.constraints.{NotNull, Size}

import com.kapilkoju.autopos.kernel.domain.AuditableEntity

import scala.beans.BeanProperty

@Entity
@Table(name = "category")
class Category extends AuditableEntity {

  @NotNull
  @Size(min = 2, max = 3)
  @Column(name = "short_name", length = 3, nullable = false)
  @BeanProperty
  var shortName: String = _

  @NotNull
  @Size(min = 3, max = 50)
  @Column(name = "name", length = 50, nullable = false)
  @BeanProperty
  var name: String = _

  @PrePersist
  @PreUpdate
  def capitalizeShortName() {
    shortName = shortName.toUpperCase
  }
}
